File: optimization-greedy-algorithms-problems-pro/src/main/java/com/greedy/problems/JobSequencing.java
package com.greedy.problems;

import com.greedy.utils.DataStructures.Job;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;

/**
 * **Problem: Job Sequencing with Deadlines**
 *
 * Given a set of jobs, where each job `i` has a deadline `d_i` and a profit `p_i`.
 * We want to find a sequence of jobs such that the total profit is maximized.
 * Each job takes one unit of time. A job can only be performed if its completion
 * time is less than or equal to its deadline. Only one job can be processed at a time.
 *
 * **Greedy Strategy:**
 * The greedy choice is to always select the job that offers the maximum profit,
 * provided it can be completed by its deadline.
 *
 * 1. Sort all jobs in descending order of their profits.
 * 2. Iterate through the sorted jobs. For each job:
 *    a. Find a time slot (from `deadline-1` down to `0`) where the job can be placed.
 *    b. If a slot is found, place the job there and mark the slot as occupied.
 *    c. Add the job's profit to the total profit.
 *
 * **Why this greedy strategy works (Intuition):**
 * **Greedy Choice Property:** By prioritizing jobs with higher profits, we ensure that
 * we always attempt to include the most valuable jobs first. If a highly profitable job `J`
 * can be scheduled, scheduling it is always beneficial. If `J` occupies a slot that could
 * have been used by a less profitable job `K`, swapping `J` for `K` (or replacing `K` with `J`
 * if `K` was already chosen) would either maintain or increase the total profit.
 * The earliest available slot strategy for a given deadline is also important: if a job
 * can be done at `t` or `t-1`, doing it at `t-1` leaves `t` open for another job,
 * potentially maximizing future options.
 *
 * **Optimal Substructure Property:** If a job `J` is chosen for a specific time slot,
 * the problem reduces to scheduling the remaining jobs in the remaining available time slots
 * to maximize their profit. The overall optimal solution is formed by combining the choice
 * for `J` with an optimal solution for the subproblem.
 *
 * **Example:**
 * Jobs: (a, 4, 20), (b, 1, 10), (c, 1, 40), (d, 1, 30)
 *
 * Sorted by Profit (descending): (c, 1, 40), (d, 1, 30), (a, 4, 20), (b, 1, 10)
 * Max Deadline = 4. Available slots: [_, _, _, _] (representing time slots 0, 1, 2, 3)
 *
 * 1. Process Job c (deadline=1, profit=40):
 *    - Try slot 0 (deadline-1): Available. Place c at slot 0.
 *    - Slots: [c, _, _, _]. Total Profit = 40.
 *
 * 2. Process Job d (deadline=1, profit=30):
 *    - Try slot 0 (deadline-1): Occupied.
 *    - No slot found <= deadline. Skip.
 *
 * 3. Process Job a (deadline=4, profit=20):
 *    - Try slot 3 (deadline-1): Available. Place a at slot 3.
 *    - Slots: [c, _, _, a]. Total Profit = 40 + 20 = 60.
 *
 * 4. Process Job b (deadline=1, profit=10):
 *    - Try slot 0 (deadline-1): Occupied.
 *    - No slot found <= deadline. Skip.
 *
 * Final Schedule: [c, _, _, a]. Max Profit = 60.
 */
public class JobSequencing {

    /**
     * Solves the Job Sequencing with Deadlines problem using a greedy approach.
     *
     * @param jobs An array of Job objects. Each job has an ID, deadline, and profit.
     * @return The maximum total profit that can be achieved.
     */
    public int findMaxProfit(Job[] jobs) {
        // Handle edge cases
        if (jobs == null || jobs.length == 0) {
            return 0;
        }

        // Step 1: Sort all jobs in descending order of their profits.
        // If profits are equal, sort by deadline ascending (optional, but can provide stable behavior).
        Arrays.sort(jobs, Collections.reverseOrder(Comparator.comparingInt(job -> job.profit)));

        // Find the maximum deadline to determine the size of the time slot array.
        int maxDeadline = 0;
        for (Job job : jobs) {
            if (job.deadline > maxDeadline) {
                maxDeadline = job.deadline;
            }
        }

        // `schedule` array will store the ID of the job scheduled at each time slot.
        // The index represents the time slot (0 to maxDeadline - 1).
        // Initialize with a sentinel value (e.g., null or a character like '0') to mark as empty.
        char[] schedule = new char[maxDeadline];
        Arrays.fill(schedule, '0'); // '0' signifies an empty slot

        int totalProfit = 0;
        List<Job> scheduledJobs = new ArrayList<>(); // To track which jobs are selected

        // Step 2 & 3: Iterate through the sorted jobs and try to schedule them.
        for (Job job : jobs) {
            // Find a free slot for this job, starting from its deadline-1 down to 0.
            // This greedy choice tries to place the job as late as possible
            // to keep earlier slots open for jobs with smaller deadlines.
            for (int i = Math.min(maxDeadline - 1, job.deadline - 1); i >= 0; i--) {
                if (schedule[i] == '0') { // If the slot is free
                    schedule[i] = job.id; // Schedule the job
                    totalProfit += job.profit;
                    scheduledJobs.add(job);
                    break; // Move to the next job
                }
            }
        }

        // Optional: Print the schedule and selected jobs
        // System.out.println("Scheduled jobs: " + scheduledJobs);
        // System.out.println("Time slots: " + Arrays.toString(schedule));

        return totalProfit;
    }

    /**
     * **Time Complexity Analysis:**
     * - Sorting jobs: O(N log N), where N is the number of jobs.
     * - Finding max deadline: O(N).
     * - Iterating through sorted jobs: O(N).
     * - Inside the loop, iterating through possible time slots: In the worst case, this inner loop
     *   can run up to `maxDeadline` times. So, O(N * maxDeadline).
     * - If `maxDeadline` is much smaller than N^2 (e.g., closer to N), the overall complexity
     *   would be closer to O(N * maxDeadline). If `maxDeadline` can be very large, this can be
     *   inefficient.
     * - Overall Time Complexity: O(N log N + N * maxDeadline). The dominant factor is usually `N log N`
     *   if `maxDeadline` is small, or `N * maxDeadline` if `maxDeadline` is large.
     *
     * **Space Complexity Analysis:**
     * - Storing jobs: O(N) for the input array.
     * - `schedule` array: O(maxDeadline) to store the job IDs for each time slot.
     * - `scheduledJobs` list: O(N) in the worst case (if all jobs are scheduled).
     * - Overall Space Complexity: O(N + maxDeadline).
     *
     * **Note on Optimizations/Variations:**
     * The slot finding (inner loop) can be optimized using a Disjoint Set Union (DSU) data structure
     * (also known as Union-Find). This allows finding the earliest free slot in nearly constant amortized time.
     * With DSU, the time complexity can be improved to O(N log N) (due to sorting) + O(N * alpha(maxDeadline))
     * where alpha is the inverse Ackermann function, which is practically a very small constant.
     * However, implementing DSU adds complexity to the code and might be overkill for a typical interview
     * unless specifically asked. The current approach is more straightforward for demonstrating the greedy logic.
     */
}

File: leetcode-greedy-algorithms-solutions-advanced/src/problems/jobSequencing.js
```javascript
/**
 * @fileoverview Implementation of the Job Sequencing with Deadlines problem using a Greedy approach.
 *
 * Problem Description:
 * Given a set of jobs, where each job has a deadline and an associated profit.
 * Each job takes unit time to complete. We want to find a sequence of jobs
 * that can be performed to maximize the total profit, given that each job
 * must be completed by its deadline.
 *
 * Example:
 * Jobs: [
 *   { id: 'a', deadline: 2, profit: 100 },
 *   { id: 'b', deadline: 1, profit: 19 },
 *   { id: 'c', deadline: 2, profit: 27 },
 *   { id: 'd', deadline: 1, profit: 25 },
 *   { id: 'e', deadline: 3, profit: 15 }
 * ]
 *
 * Optimal Solution:
 * The greedy strategy would typically select jobs 'a', 'c', 'e' for a total profit of 142.
 * Job 'a' (profit 100) has deadline 2.
 * Job 'c' (profit 27) has deadline 2.
 * Job 'e' (profit 15) has deadline 3.
 *
 * Schedule:
 * Time Slot 1: Job 'a' (due to higher profit than 'd' and 'b' for deadline 1 slot)
 * Time Slot 2: Job 'c' (due to higher profit than 'b' for deadline 2 slot)
 * Time Slot 3: Job 'e'
 *
 * Note: If 'd' had a higher profit, e.g., 150, then 'd' would be chosen for slot 1,
 * and 'a' would be missed or pushed if possible (but here 'a' can't be pushed further than deadline 2).
 *
 * Greedy Choice Property:
 * The greedy strategy is to process jobs in decreasing order of their profits.
 * For each job, try to schedule it in the latest possible time slot (from 1 to its deadline)
 * that is still available. This is because scheduling a high-profit job later keeps
 * earlier slots free for other high-profit jobs with earlier deadlines.
 *
 * Proof of Correctness (Intuition):
 * Assume there is an optimal schedule `S` that does not follow the greedy strategy.
 * This means there exist two jobs `i` and `j` such that `profit[i] > profit[j]`,
 * but `j` is scheduled and `i` is either not scheduled or scheduled at an earlier slot
 * that could have been used by `i` (assuming `i` meets its deadline).
 *
 * If `j` is scheduled at time `t` and `i` is not scheduled, but `i` could have been
 * scheduled at `t'` (where `t' <= deadline[i]`) and `t'` was free (or occupied by a
 * lower-profit job than `i` that could be moved), then we can swap `i` and `j`
 * (or schedule `i` in `t'`) to achieve a higher or equal profit without violating deadlines.
 *
 * More formally, by sorting jobs by profit in descending order, we ensure that
 * we prioritize higher-profit jobs. When considering a job, we try to place it
 * in the latest possible free slot up to its deadline. This maximizes the chances
 * for other profitable jobs (which might have tighter deadlines) to be scheduled
 * in earlier slots. If we place a job `J` with profit `P` at time `t`, and `t` is the
 * latest possible free slot before `J.deadline`, this decision is locally optimal.
 * If there was another job `J'` with a lower profit `P'` scheduled at `t'`, and `J`
 * could have been scheduled at `t'`, but `J'` could not have been scheduled at `t`,
 * then swapping them would not decrease total profit.
 */

/**
 * Solves the Job Sequencing with Deadlines problem to maximize total profit.
 *
 * @param {Array<Object>} jobs - An array of job objects.
 *   Each job object must have 'id' (string), 'deadline' (number), and 'profit' (number).
 *   Example: [{ id: 'a', deadline: 2, profit: 100 }, { id: 'b', deadline: 1, profit: 19 }]
 * @returns {Object} An object containing:
 *   - {number} totalProfit: The maximum total profit achievable.
 *   - {Array<string>} sequencedJobs: An array of job IDs representing the optimal sequence.
 *
 * Time Complexity: O(N log N) due to sorting, plus O(N * MaxDeadline) for scheduling
 *                  in the worst case (iterating up to MaxDeadline for each job).
 *                  If MaxDeadline is large, this can approach O(N*N).
 *                  A more optimized approach using a Disjoint Set Union (DSU) or Max-Heap
 *                  can reduce the scheduling part to O(N log MaxDeadline) or O(N log N).
 *                  For typical interview constraints where MaxDeadline isn't excessively large,
 *                  the current array-based approach is often accepted.
 * Space Complexity: O(MaxDeadline) for the `slots` array, plus O(N) for sorted jobs and result.
 */
function jobSequencing(jobs) {
  // Edge case: No jobs or invalid input.
  if (!jobs || jobs.length === 0) {
    return { totalProfit: 0, sequencedJobs: [] };
  }

  // 1. Sort jobs by profit in descending order.
  // This is the primary greedy choice. High-profit jobs are considered first.
  const sortedJobs = [...jobs].sort((a, b) => b.profit - a.profit);

  // Find the maximum deadline to determine the size of our time slot array.
  const maxDeadline = Math.max(...jobs.map(job => job.deadline));

  // Initialize an array to keep track of available time slots.
  // `slots[i]` will store the ID of the job scheduled in time slot `i+1`.
  // Using `null` or `undefined` to denote an empty slot.
  // Slots are 1-indexed for deadlines, so array size is maxDeadline.
  const slots = new Array(maxDeadline).fill(null);

  let totalProfit = 0;
  const sequencedJobsIds = [];

  // 2. Iterate through the sorted jobs.
  for (const job of sortedJobs) {
    // Try to place the current job into the latest possible free slot
    // before or at its deadline.
    // We iterate backwards from `job.deadline - 1` down to `0` (representing slot 1).
    for (let i = Math.min(job.deadline - 1, maxDeadline - 1); i >= 0; i--) {
      if (slots[i] === null) {
        // Slot `i+1` is free, schedule this job.
        slots[i] = job.id;
        totalProfit += job.profit;
        sequencedJobsIds.push(job.id);
        break; // Move to the next job, this one is scheduled.
      }
    }
  }

  // Jobs were added in order of processing (highest profit first), but
  // the actual sequence of execution is determined by their assigned slots.
  // Filter out nulls and map to job IDs to get the chronological sequence.
  const finalSequencedJobs = slots.filter(jobId => jobId !== null);

  return { totalProfit, sequencedJobs: finalSequencedJobs };
}

export default jobSequencing;
```